CIP-0157? | Cardano URIs - Enhanced Payments

Crypto2099 opened the pull request:
[Rendered Version](https://github.com/Crypto2099/CIPs/tree/cardano-uris-enhanced-payment/CIP-XXXX)

This proposal seeks to extend the CIP-13 payment URIs and address some of the issues raised here: https://github.com/cardano-foundation/CIPs/issues/836

We need an enhanced and more explicit `Payment` type of Cardano URI in the modern era of native assets, smart contracts, and transactional metadata.

Because of legacy integrations with existing URI payments (lacking a dedicated authority) I have chosen to break this extension out into a new CIP with an explicit authority to avoid having to change or update CIP-13.

Crypto2099 committed:
Initial Commit

rphair commented:
@Crypto2099 good to have a stub here for the current discussion of how a comprehensive and collision-free payment URI syntax would look.  Cross-referencing my https://github.com/cardano-foundation/CIPs/issues/836#issuecomment-2156442263 and https://github.com/cardano-foundation/CIPs/issues/836#issuecomment-2147763143 to be sure scenarios involving these boundary cases will be handled properly by any proposed syntax:
- keyword(s) used to specify metadata collide with token name(s)

... including these cases if payment handles are supported by this protocol (I would insist that they should be):
- payment handles (if supported by this protocol) are ambiguous as per https://github.com/cardano-foundation/CIPs/pull/605 (cc @SIDANWhatever)
- keyword(s) used to disambiguate payment handles collide with token name(s)
- payment handles collide with bare payment addresses (unlikely but possible)

Crypto2099 commented:
Thanks @rphair excited to start the conversation and hoping to have some first-draft specs delivered over the course of the next week for discussion! These are great starting points for consideration though!

rphair commented:
you're welcome @Crypto2099 & if you feel potential specs could use community brainstorming then feel free to add it to the agenda in a week-plus time: https://hackmd.io/@cip-editors/91

MadOrkestra commented:
Quick heads up: beginWallet implemented a "CIP13-like" solution now (https://x.com/BeginWallet/status/1941614893691724100) - so I'll leave this here as starting point for a discussion to turn this into a standard CIP addressing at least _some_ of the downfalls of the original payment scheme.

the beginWallet URI scheme @francisluz already implemented looks like this:
`web+cardano:addr?amount=1&token=ADA&paymentID=123-ABC&note=abcdedafad`

**What it solves**
- Payments with other tokens than ADA (think USDM)
- Linking payments to offchain data (to link a payment to a shopping cart of whatever)

**What it doesn't solve / Downfalls**
- Transactions with multiple tokens
- Uses a ticker symbol instead of the policy id, which then has to be verified/looked up by a centralized whitelist in each wallet
- Not using the //payment authority

I think the biggest issue here is the ticker symbol and the centralized whitelists that have to come with it to convert the ticker into a policyId to build the transaction. But: I think we have to abandon the original idea of multiple assets to fit into a payment URL anyways, because while URLs can be quite lengthy today in most modern browsers (up to 10,000 characters or something), the bottleneck is the QR-Code with "only" 300-500 characters before the "scanability" decreases massively, which especially for the use case of providing a slick and quick payment method is a no-go.

If we were to go with policyIds instead of ticker symbols, the overall character length excluding the note/msg field would still perfectly fit within these limitations: 
- Identifier, params and stuff: 43+ Characters
- Amount: 1 to whatever (100bn are still only 12 digits)
- Bech32 Address: 104 Characters 
- PolicyId: 56 Characters
- PaymentID: 32-36 Characters (for UUIDv4, could be shorter ofc)

I think the dedicated paymentId field is debatable, as long as you can put _some_ metadata into a transaction via URL, we cover a lot of use cases.
 
Given that we've spent another year or so without proper definition of this very rudimentary functionality on Cardano, I'd rather see this moving quickly now and enabling stable coins or Hosky or whatever as possible payment method across the ecosystem instead of looking for the perfect solution while seeing wallets coming up with their own individual solutions, but thats just my take ofc.

rphair commented:
thanks @MadOrkestra - this covers everything I would have said.  Just one clarification though (cc @francisluz):
> **What it solves**
> - Payments with other tokens than ADA

When paying a non-Ada token, is the minimum amount of ada (to carry the tokens) implicit in the transfer?

Crypto2099 committed:
Merge branch 'cardano-foundation:master' into cardano-uris-enhanced-payment

Crypto2099 committed:
Updates to finally add the proper specification and bring the standard to the masses

rphair reviewed (commented):
Looking forward to confirming this at the next meeting (https://hackmd.io/@cip-editors/116) and recording the feedback here from any interested parties especially @MadOrkestra @francisluz and any other implementors following this issue.

I might only be half-complete tagging reps from other wallets besides Begin, so please I will leave that to @MadOrkestra @Crypto2099 @Ryun1 - with invitations to that meeting & calls for review and suggestions all around this proposal.  (Normally we would wait 1 more meeting cycle for detailed Review after Triage first, but time seems essential because of the implementation in progress.)

rphair reviewed (commented):
Great work finally bringing this home @Crypto2099 - candidacy confirmed @ yesterday's CIP meeting & please update the document & directory name accordingly 🎉

Personally I expect to ✅ this soon but want to give it one more detailed read when settled again from travelling around the end of this week: especially cross referencing with CIP-0013 family members & #1058.

MadOrkestra commented:
Way more complex than I initially thought, but great job @Crypto2099 - I have one thing to at least think about: As per [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986) "|" (used to separate the asset count here) is not allowed as unencoded character in URLs. Not sure if this can cause problems with individual URL schemes as web+cardano, but it might be worth looking into?

Crypto2099 committed:
Update CIP Number

Co-authored-by: Robert Phair <rphair@cosd.com>

MadOrkestra commented:
> > I have one thing to at least think about: As per [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986) "|" (used to separate the asset count here) is not allowed as unencoded character in URLs. Not sure if this can cause problems with individual URL schemes like web+cardano, but it might be worth looking into?
> 
> Thanks for the heads up on this, I will take a look and pivot the standard to use a URI-safe character

Sorry, didn't have too much time for this earlier... my understanding is, it it basically comes down to one of these: 

`- . _ ~ ! $ ( ) * + ;`

So unless we wanna drive every single dev out there mad with bringing back the good old tilde or cause confusion with a semicolon, I think `?t=asset1*amount1,asset2*amount2` is the logical way to go? Amounts could be optional and have a decimal point too if needed with no URL encoding at all.

Crypto2099 commented:
> > > I have one thing to at least think about: As per [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986) "|" (used to separate the asset count here) is not allowed as unencoded character in URLs. Not sure if this can cause problems with individual URL schemes like web+cardano, but it might be worth looking into?
> > 
> > 
> > Thanks for the heads up on this, I will take a look and pivot the standard to use a URI-safe character
> 
> Sorry, didn't have too much time for this earlier... my understanding of RFC 3986 is, it it basically comes down to one of these:
> 
> `- . _ ~ ! $ ( ) * + ;`
> 
> So unless we wanna drive every single dev out there mad with bringing back the good old tilde or cause confusion with a semicolon, I think `?t=asset1*amount1,asset2*amount2` is the logical way to go? Amounts could be optional (1 if not set - to save URL characters) and have a decimal point too (if needed) - with no URL encoding necessary at all.

The asterisk should do the job nicely without causing too much confusion (like the dash or plus sign would). I think we can stick to whole-integer representation since the ledger does not natively understand Native Asset decimals anyway and wallets need to build the transaction with the whole integer amount of the native asset in question. I do agree that removing `*1` also seems a logical default to save characters and space.

Crypto2099 committed:
Updates based on feedback
- Added initial version information to URI
- Expanded abbreviated addresses and asset IDs
- Clarified some info on token parameters such as whole integer quantities and making quantity default to 1 if not otherwise specified
- Switch from pipe character to the use of an asterisk for the quantity separator for RFC compliance

AlexDochioiu commented:
Appreciate the update to support asset unit @Crypto2099 🙏🏻

AlexDochioiu commented:
@Crypto2099 I'd add an explicit remark in the CIP that token quantity **is not** decimals adjusted. (token quantity is always `int` / `bigint` ; never `float` or `decimal`)

Kingsleyioryoo reviewed (commented):
1DdfpmAw7QKogrSQ2VcDhqjtoXyHHPQq6f

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,301 @@
+### Acceptance Criteria
+
+* [ ] Community Feedback and Review Integrated
+* [ ] At least one wallet supports this payment standard
+* [ ] At least one project using this payment standard
```
```suggestion
* [ ] At least one wallet supports this payment standard
* [ ] At least one project using this payment standard
```
As I also posted in https://github.com/cardano-foundation/CIPs/pull/1058#discussion_r2239029563 @Ryun1 (our usual parameter setter for this, or @perturbing or any other community reviewer) may think this should be increased from 1 to 2 or maybe more.

Personally I think 1 is fine on each side (producer + consumer) due to reasons of inevitable imitation that I mentioned in the other comment: but will agree to increase it if other editors request.

@Crypto2099 I do think the term `project` needs to be elaborated, since wallets are also "projects" — whatever it takes to establish that the "producer" and "consumer" of payment URIs are distinct projects.  As we mentioned at the meeting I think it would also help to specify that those 2 projects should be from distinct _vendors_... a statement that would also support the `Path to Active` credibility of https://github.com/cardano-foundation/CIPs/pull/1058.

rphair reviewed (commented):
As I just posted for candidate CIP-0158 (Browse Authority) (https://github.com/cardano-foundation/CIPs/pull/1058#pullrequestreview-3066295653) I believe this has all the elements of a proper CIP: with exhaustive detail & pre-emptive consideration of valuable use cases.

I think it is especially valuable that an ABNF grammar has been included: not because it will be automatically parseable or useful in validating URI syntax, but to eliminate ambiguities & edge cases that could otherwise emerge between different wallets & apps.

So as also posted for the parallel candidate CIP I am tentatively ✅ing this proposal as we begin the inevitable period of convergence between in-progress implementations (thanks @AlexDochioiu) and the standard expressed in ideal form.

If implementors ever think it would help to brainstorm in a "working group" setting around particular syntax / interpretation challenges for either of these proposals, please post to say so here and I'll put it up for Review on the CIP agenda.

rphair commented:
@Crypto2099 you're the best candidate I can think of to bring together all possible approaches to this... so please let us know what you think needs to be done & editors will do anything possible to ensure it's merged promptly: with no time to waste, as @MadOrkestra says 😅

colll78 commented on CIP-XXXX/README.md:
````diff
@@ -0,0 +1,301 @@
+* 5x of asset17q7r59zlc3dgw0venc80pdv566q6yguw03f0d9
+
+  ```
+  t=asset12ffdj8kk2w485sr7a5ekmjjdyecz8ps2cm5zed*100,asset17q7r59zlc3dgw0venc80pdv566q6yguw03f0d9*5
````
I would recommend that CIP-157 encode the full ledger asset ID instead; either policyId.assetNameHex or plain policyIdAssetNameHex. 

A CIP-14 fingerprint is a 20-byte hash of policyId || assetName. It is not a reversible encoding, so wallets cannot turn it into the ledger unit they need to build the corresponding tx.

colll78 replied:
The issues I see:
- CIP-14 fingerprints are not reversible.
- The grammar requires *quantity, while the accompanying note says quantity is optional and defaults to one.
- The on-chain meaning of i= and n= is unspecified.
- Duplicate and unknown query-parameter behavior is unspecified.
- “64 characters” conflicts with the ledger metadata constraint, which is 64 UTF-8 bytes per text value.
- Zero is permitted by 1*DIGIT, despite the prose requiring positive quantities.

rphair replied:
@francisluz how would you feel about changing your implementation to match what @colll78 is suggesting?  Please <ins>we need you to answer this</ins> because, at this point, this CIP draft should either be merged or discarded in favour of whatever can be written to match future implementations.  

So far Begin Wallet has the only credit for an implementation, while @AlexDochioiu has said in https://github.com/cardano-foundation/CIPs/issues/836#issuecomment-3961657610 that VESPR might support this someday.  Failing to merge this CIP because of lack of confirmation would invalidate this tentative progress: but would avoid the dev community wondering whether something called "CIP-0157" needs to be implemented and stalling any further attempts to standardise payment links (cc @Crypto2099).

If using the asset ID is the only way to future-proof this, then we need to make this change so other wallet devs can implement this in confidence.  If this is the _only_ issue @colll78 is bringing up in review, then I trust that doing it this way is important even if breaking some already issued links.

So please @francisluz if you think that Begin Wallet users (or any other Cardano payments) depend upon the original format then maybe this new suggestion can be written into a `v2` as encoded in the URL.  Again, I would rather see a general standard for payment links begin with Version 2 than never begin at all.

rphair committed:
Merge branch 'master' into cardano-uris-enhanced-payment

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,301 @@
+    - Begin Wallet <@francisluz>
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/843
+    - https://github.com/cardano-foundation/CIPs/issues/836
```
```suggestion
    - Original PR: https://github.com/cardano-foundation/CIPs/pull/843
    - Implementation discussion: https://github.com/cardano-foundation/CIPs/issues/836
```
New format (& workflow validation) requires labels for links (brief & descriptive by editorial standard).

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,301 @@
+_Metadata_ as well as providing for extensibility and versioning. All features
+lacking in the original [CIP-13] payment URIs.
+
+## Motivation: why is this CIP necessary?
```
```suggestion
## Motivation: Why is this CIP necessary?
```
(1 of 2) fix long standing capitalisation glitch in standard template

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,301 @@
+Supports backward compatibility for Byron-era Base58 addresses. All query
+parameters work the same.
+
+## Rationale: how does this CIP achieve its goals?
```
```suggestion
## Rationale: How does this CIP achieve its goals?
```
(2 of 2) fix long standing capitalisation glitch in standard template

rphair reviewed (commented):
Given recent discussion of how payment links might accommodate "handles":
* https://github.com/cardano-foundation/CIPs/pull/1229#discussion_r3742436344

— and because we already have one implementing wallet (Begin) and another [willing to implement](https://github.com/cardano-foundation/CIPs/issues/836#issuecomment-3961657610) (VESPR, as per @AlexDochioiu) —  we should have this CIP confirmed by merging as our best effort to standardise payment links.  I've made some arguments for this in the course of our only recommended technical change so far (https://github.com/cardano-foundation/CIPs/pull/843#discussion_r3745455357).

We should also keep in mind the version currently `/v1` in the URL can always be bumped if mutual compatibility isn't perfect the first time, or needs to specify additional things like "handle resolvers" (cc @kingslavcho @papag00se).

Given @colll78's latest https://github.com/cardano-foundation/CIPs/pull/843#discussion_r3743952708 I would recommend this course:
1. a rewrite as he suggests... or however that discussion concludes (and hopefully he or @francisluz would please submit this as a correction & editors will commit it);
2. I'm submitting a format edit below to match our current CIP standards & pass validation (vs. [these failures](https://github.com/cardano-foundation/CIPs/actions/runs/31337336961/job/93305240387?pr=843));
3. once all edits are in, I would approve this & tag `Last Check` to get this merged ASAP — especially since it would allow further deltas to be submitted freely: with appreciation for how far @Crypto2099 has carried this already (cc @Ryun1 @perturbing);
4. if @Crypto2099 cannot rename the containing directory to `CIP-0157` as suggested long ago, the editors can fix the repository structure accordingly after this is merged.

rphair committed:
validation thing: add labels to `Discussions:`

rphair committed:
(1 of 2) validation thing: canonical section title

rphair committed:
(2 of 2) validation thing: canonical section title
